api/search: Reset `is_automated` to false

vvoland opened the pull request:
- related to: https://github.com/moby/moby/pull/45925

The field will still be present in the response, but will always be `false`.
Searching for `is-automated=true` will yield no results, while `is-automated=false` will effectively be a no-op.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

**- How to verify it**
CI

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
The `is_automated` field in the `POST /images/search` endpoint results is always `false` now. Consequently, searching for `is-automated=true` will yield no results, while `is-automated=false` will be a no-op.
```

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -489,6 +489,15 @@ func (ir *imageRouter) getImagesSearch(ctx context.Context, w http.ResponseWrite
 		return err
 	}
 
+	// "is-automated" is deprecated and was removed in 1.45.
+	// If the client is older than 1.45, remove the filter to avoid erroring out
+	// on older clients.
+	if versions.LessThan(httputils.VersionFromContext(ctx), "1.45") {
+		for _, v := range searchFilters.Get("is-automated") {
+			searchFilters.Del("is-automated", v)
+		}
+	}
+
```
Following our discussion on Slack; I think the plan was to keep the filter, but it would not produce results; from the original PR;

> Consequently, searching for `is-automated=true` will yield no results.

And the change to have `IsAutomated` always being `false` would be unversioned;

> The deprecation is not versioned, and applies to all API versions.

Motivation there was that the deprecation correlates with the upstream (docker hub) service being in process of deprecating it; IIRC, the field is already out of date with reality, so results are not correct in many cases, thus always setting it to `false` makes sure that we don't return "bogus" results here.

vvoland replied:
Yeah so with this PR the `IsAutomated` response will always be `false` (regardless of API version), but the `is-automated` filter will always produce an error on API >= 1.45 (it no longer makes sense, so we don't want newer clients to be using it while still keeping old clients working).

thaJeztah commented on api/swagger.yaml:
```diff
@@ -8768,16 +8768,6 @@ paths:
-
-                    <p><br /></p>
-
-                    > **Deprecated**: This field is deprecated and will always
-                    > be "false" in future.
-                  type: "boolean"
-                  example: false
```
We can keep the field (removing it was considered more problematic for clients that don't "unmarshal" but directly use the response), but it would always be "false"

vvoland replied:
Ah yeah, deleted too much here. Restored this one.

corhere commented on docs/api/version-history.md:
```diff
@@ -19,6 +19,9 @@ keywords: "API, Docker, rcli, REST, documentation"
   subpath of a named volume to be mounted.
+* Removed the `is-automated` filter on the `POST /images/search` endpoint.
+  The `is_automated` field has been deprecated by Docker Hub's search API.
+  Consequently, searching for `is-automated=true` will yield no results.
```
```suggestion
  Consequently, searching for `is-automated=true` or `is-automated=false`
  will yield an "invalid filter" error.
```

corhere commented on registry/search.go:
```diff
@@ -74,6 +62,8 @@ func (s *Service) Search(ctx context.Context, searchFilters filters.Args, term s
 				continue
 			}
 		}
+		// "is-automated" is deprecated and will always be false
+		result.IsAutomated = false //nolint:staticcheck  // ignore SA1019 (field is deprecated)
```
If it's deprecated and always false in Hub results, why do we need to explicitly set it to false ourselves?

vvoland replied:
IIUC, on Hub side this is not always false (yet), it's just deprecated.

But @thaJeztah  might have more context here

corhere commented on registry/search_test.go:
```diff
@@ -97,15 +97,8 @@ func TestSearchErrors(t *testing.T) {
-			expectedError: "invalid filter 'is-automated",
+			filtersArgs:   filters.NewArgs(filters.Arg("is-automated", "true")),
+			expectedError: "invalid filter 'is-automated'",
 		},
```
```suggestion
		},
		{
			filtersArgs:   filters.NewArgs(filters.Arg("is-automated", "false")),
			expectedError: "invalid filter 'is-automated'",
		},
```

corhere reviewed (commented):
If we're going to make `is-automated=false` an invalid filter on v1.45+ we should be upfront about that.

thaJeztah commented:
I don't think we need to remove the filter, and just silently ignoring would be fine.

The reason I'd want the field to be reset is that it's very much broken. The v1 search is a bit of a legacy service, and this field is known to be broken. In all honesty, it was announced internally to deprecated that field (as repairing would not be possible), but that task has been kicked around to different teams, so I have no clue when that work continues, other than knowing it's broken, and that it won't be fixed.

vvoland commented:
> I don't think we need to remove the filter, and just silently ignoring would be fine.

What's the point of having that filter on `>= 1.45` though? If it's going to be a no-op anyway, we should just get rid of it any only ignore it for older APIs.

corhere commented:
> What's the point of having that filter on `>= 1.45` though?

[To not break users' existing workflows.](https://devblogs.microsoft.com/oldnewthing/20240216-00/?p=109409)

vvoland commented:
And it won't break if they pin to an API version before 1.45, where it will just be silently ignored:

https://github.com/moby/moby/blob/b1039b56b19196fc8c2822a8897b76b90bf7d53c/api/server/router/image/image_routes.go#L492-L499

vvoland committed:
api/search: Reset `is_automated` field to `false`

The field will still be present in the response, but will always be
`false`.
Searching for `is-automated=true` will yield no results, while
`is-automated=false` will effectively be a no-op.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
As discussed on maintainers call, updated the PR to not remove the filter completely.

thaJeztah reviewed (approved):
LGTM

corhere commented on docs/api/version-history.md:
```diff
@@ -19,6 +19,9 @@ keywords: "API, Docker, rcli, REST, documentation"
   subpath of a named volume to be mounted.
+* `POST /images/search` will always assume a `false` value for the `is-automated`
+  field. Consequently, searching for `is-automated=true` will yield no results,
+  while `is-automated=false` is will be a no-op.
```
```suggestion
  while `is-automated=false` will be a no-op.
```

vvoland replied:
🤦🏻 Thanks, fixed!

corhere commented on registry/search.go:
```diff
@@ -74,6 +73,8 @@ func (s *Service) Search(ctx context.Context, searchFilters filters.Args, term s
 				continue
 			}
 		}
+		// "is-automated" is deprecated and will always be false
```
```suggestion
		// "is-automated" is deprecated and the value in Docker Hub search results is untrustworthy. Force it to false so as to not mislead our clients.
```

thaJeztah reviewed (approved):
still LGTM, thanks!
